### PR TITLE
[Judges] use the pair-judges in online-preference trainers

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

add support for judges API in online-pref methods

### CODE REVIEW

File: trl/trainer/online_dpo_trainer.py [hunk 1]
```diff
@@ -419,23 +419,35 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
...
 
-        # Filter completion. Ensure that the sample contains stop_token_id
-        # Completions not passing that filter will receive a lower score.
-        contain_eos_token = torch.any(completion_ids == self.processing_class.eos_token_id, dim=-1)
-        if self.args.missing_eos_penalty is not None:
-            scores[~contain_eos_token] -= self.args.missing_eos_penalty
+            # Filter completion. Ensure that the sample contains stop_token_id
+            # Completions not passing that filter will receive a lower score.
+            contain_eos_token = torch.any(completion_ids == self.processing_class.eos_token_id, dim=-1)
+            if self.args.missing_eos_penalty is not None:
```
Comment (qgallouedec) [inline trl/trainer/online_dpo_trainer.py:441]: maybe add a warning when `args.missing_eos_penalty` is not `None` and `judge`not `None`

Comment (qgallouedec) [reply trl/trainer/online_dpo_trainer.py:441]: in the init

File: trl/trainer/online_dpo_trainer.py [hunk 2]
```diff
@@ -419,23 +419,35 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
...
+        # Get the reward from the reward model or judge:
+        if self.judge is not None:
+            prompts = [self.processing_class.decode(ids[:context_length]) for ids in prompt_completion_ids]
+            completions = [self.processing_class.decode(ids[context_length:]) for ids in prompt_completion_ids]
+            ranks_of_first_completion = self.judge.judge(
+                prompts, [[completions[i], completions[i + num_examples]] for i in range(num_examples)]
+            )
+            # convert ranks to a True/False mask:
+            # when rank == 0 it means the first completion is the better
+            # when rank == 1 it means the second completion is the bette
```
Comment (qgallouedec) [inline trl/trainer/online_dpo_trainer.py:431]:
```suggestion
            # when rank == 1 it means the second completion is the better
```

It can also be -1 right?

File: trl/trainer/online_dpo_trainer.py [hunk 3]
```diff
@@ -419,23 +419,35 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
...
-            # Get the reward from the reward model
+        # Get the reward from the reward model or judge:
+        if self.judge is not None:
+            prompts = [self.processing_class.decode(ids[:context_length]) for ids in prompt_completion_ids]
+            completions = [self.processing_class.decode(ids[context_length:]) for ids in prompt_completion_ids]
+            ranks_of_first_completion = self.judge.judge(
+                prompts, [[completions[i], completions[i + num_examples]] for i in range(num_examples)]
+            )
+            # convert ranks to a True/False mask:
+            # when rank == 0 it means the first completion is the better
```
Comment (qgallouedec) [inline trl/trainer/online_dpo_trainer.py:430]:
```suggestion
            # when rank == 0, it means that the first completion is the best
```

**Comment by qgallouedec** [comment]
@kashif can you also add an example in the online dpo documentation? And a test?

**Comment by kashif** [comment]
working on test thanks!

File: trl/trainer/online_dpo_trainer.py [hunk 4]
```diff
@@ -419,23 +421,41 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
             ref_logprobs = torch.take_along_dim(ref_all_logprobs, completion_ids.unsqueeze(-1), dim=2).squeeze(-1)
             del ref_output, ref_logits, ref_all_logprobs  # free memory
 
-            # Get the reward from the reward model
+        # Get the reward from the reward model or judge:
+        if self.judge is not None:
+            prompts = [
+                self.processing_class.decode(ids[:context_length], skip_special_tokens=True)
+                for ids in prompt_completion_ids[:num_examples]
+            ]
```
Comment (qgallouedec) [inline trl/trainer/online_dpo_trainer.py:429]: can we use `batch_decode` here?

File: trl/trainer/nash_md_trainer.py [hunk 5]
```diff
@@ -354,28 +374,28 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
...
+        # self._log_statistics(
+        #     model_data,
+        #     mixture_data,
+        #     model_logprobs_model_data.detach(),
+        #     ref_logprobs_model_data,
+        #     probability,
+        #     score.detach(),
+        #     kl_div.detach(),
+        #     context_length,
+        # )
```
Comment (qgallouedec) [inline trl/trainer/nash_md_trainer.py:398]: wip?

File: trl/trainer/online_dpo_trainer.py [hunk 6]
```diff
@@ -156,12 +156,13 @@ def __init__(
             )
```
Comment (qgallouedec) [inline trl/trainer/online_dpo_trainer.py:156]:
```suggestion
            )
            judge = None
```

File: trl/trainer/online_dpo_trainer.py [hunk 7]
```diff
@@ -469,22 +489,25 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
...
+
         kl = logprobs - ref_logprobs
         mean_kl = kl.sum(1).mean()
         self.stats["objective/kl"].append(self.accelerator.gather(mean_kl).mean().item())
         non_score_reward = (-self.beta * kl).sum(1)
         mean_non_score_reward = non_score_reward.mean()
         self.stats["objective/non_score_reward"].append(self.accelerator.gather(mean_non_score_reward).mean().item())
-        rlhf_reward = scores + non_score_reward
-        self.stats["objective/rlhf_reward"].append(self.accelerator.gather(rlhf_reward).mean().item())
+        if self.judge is None:
```
Comment (qgallouedec) [inline trl/trainer/online_dpo_trainer.py:506]: Equivalent, but clearer this way imo
```suggestion
        if self.reward_model is not None:
```

File: trl/trainer/online_dpo_trainer.py [hunk 8]
```diff
@@ -469,22 +489,25 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
         loss = losses.mean()
 
         # Log everything
+        if self.judge is None:
```
Comment (qgallouedec) [inline trl/trainer/online_dpo_trainer.py:492]:
```suggestion
        if self.reward_model is not None:
```

File: trl/trainer/xpo_trainer.py [hunk 9]
```diff
@@ -135,9 +138,6 @@ def __init__(
             "objective/kl": [],
             "objective/entropy": [],
             # Replace "scores" by "model_scores" and "ref_scores"
-            "objective/model_scores": [],
-            "objective/ref_scores": [],
-            "objective/scores_margin": [],
```
Comment (qgallouedec) [inline trl/trainer/xpo_trainer.py:140]: we don't log the score anymore?

File: tests/test_online_dpo_trainer.py [hunk 10]
```diff
@@ -198,3 +198,30 @@ def test_training_with_peft_model_and_peft_config(self):
...
+                learning_rate=5.0e-7,
+                eval_strategy="steps",
+                report_to="none",
+            )
+            dummy_dataset = load_dataset("trl-internal-testing/zen", "standard_prompt_only")
+
+            trainer = OnlineDPOTrainer(
+                model=self.model,
+                judge=PairRMJudge(),
+                reward_model=None,
```
Comment (qgallouedec) [inline tests/test_online_dpo_trainer.py:218]:
```suggestion
```

**Comment by qgallouedec** [comment]
```
# 8 GPUs
accelerate launch examples/scripts/dpo_online.py \
    --model_name_or_path trl-lib/pythia-1b-deduped-tldr-sft  \
    --judge pairrm \
    --dataset_name trl-lib/tldr \
    --learning_rate 5.0e-7 \
    --output_dir pythia-1b-tldr-online-dpo-reward \
    --warmup_ratio 0.1
```

https://wandb.ai/huggingface/huggingface/runs/1meibbsh

File: trl/trainer/online_dpo_config.py [hunk 11]
```diff
@@ -32,7 +32,9 @@ class OnlineDPOConfig(TrainingArguments):
             Initial learning rate for [`AdamW`] optimizer. The default value replaces that of
             [`~transformers.TrainingArguments`].
         reward_model_path (`Optional[str]`, *optional*, defaults to `None`):
-            Path to the reward model.
+            Path to the reward model. Either `judge` or `reward_model_path` must be set, but not both.
+        judge (`Optional[str]`, *optional*, defaults to `None`):
```
Comment (lewtun) [inline trl/trainer/online_dpo_config.py:36]: Should we raise an error in the post init of these configs if both are set?

File: examples/scripts/dpo_online.py [hunk 12]
```diff
@@ -81,12 +85,21 @@
         model_config.model_name_or_path, trust_remote_code=model_config.trust_remote_code, **model_kwargs
     )
 
-    reward_model = AutoModelForSequenceClassification.from_pretrained(
-        training_args.reward_model_path,
-        num_labels=1,
-        trust_remote_code=model_config.trust_remote_code,
-        **model_kwargs,
-    )
+    if training_args.reward_model_path is not None:
```
Comment (lewtun) [inline examples/scripts/dpo_online.py:88]: WDYT about having a helper method like `get_reward_model(TrainingArguments, ModelConfig)` so we can (a) unit test it and (b) not have duplicate logic across all scripts?

File: examples/scripts/dpo_online.py [hunk 13]
```diff
@@ -57,6 +59,8 @@
 from trl.trainer.utils import SIMPLE_CHAT_TEMPLATE
 
 
+JUDGES = {"pair_rm": PairRMJudge, "openai": OpenAIPairwiseJudge}
```
Comment (lewtun) [inline examples/scripts/dpo_online.py:62]: Should we include HF ones too?
```suggestion
JUDGES = {"pair_rm": PairRMJudge, "openai": OpenAIPairwiseJudge, "hf": HfPairwiseJudge}
```

File: trl/trainer/xpo_trainer.py [hunk 14]
```diff
@@ -220,6 +223,25 @@ def _compute_rewards(self, model_data, ref_data, context_length):
 
         return model_scores, ref_scores
 
+    def _compute_judge(self, model_data, ref_data, context_length):
+        prompts = self.processing_class.batch_decode(
```
Comment (lewtun) [inline trl/trainer/xpo_trainer.py:227]: Is skipping special tokens a good idea for chat templates like ChatML? Otherwise I think we'll lose tokens like `<|im_start|>` and `<|im_end|>` which may confuse the judge

**Review by lewtun — COMMENTED** [review]
Thanks for adding this neat feature @kashif ! I left some questions, but otherwise this is looking pretty good 🚀

File: trl/trainer/xpo_trainer.py (hunk 14 above)
Comment (kashif) [reply trl/trainer/xpo_trainer.py:227]: right but my feeling was that models like pair-rm take in sentences as input... I can try to add the tokens back but then we also get padding tokens etc. which would need removing as well?

File: examples/scripts/dpo_online.py (hunk 13 above)
Comment (qgallouedec) [reply examples/scripts/dpo_online.py:62]: LGTM. Of course, if a user needs to use these judges with parameters other than the default, they will have to modify the script. This is the case, for example, if they want to use a model other than Meta-Llama-3-70B-Instruct for `HfPairwiseJudge`.

File: trl/trainer/xpo_trainer.py (hunk 14 above)
Comment (qgallouedec) [reply trl/trainer/xpo_trainer.py:227]: That's tricky. For some tokenizers, as @lewtun says, when you skip the special tokens, it seems that it's not a conversation:

```python
from transformers import AutoTokenizer

messages = [
    [
        {"role": "user", "content": "What color is the sky?"},
        {"role": "assistant", "content": "The sky is blue."},
    ],
    [
        {"role": "user", "content": "What is the closest planet to the sun?"},
        {"role": "assistant", "content": "It is Mercury."},
    ],
]

tokenizer = AutoTokenizer.from_pretrained("trl-internal-testing/tiny-random-Mistral-7B-Instruct-v0.3")
tokenizer.pad_token = tokenizer.eos_token
input_ids = tokenizer.apply_chat_template(messages, padding=True)

print(tokenizer.batch_decode(input_ids, skip_special_tokens=True))
['What color is the sky? The sky is blue.', 'What is the closest planet to the sun? It is Mercury.']
print(tokenizer.batch_decode(input_ids, skip_special_tokens=False))
['</s></s></s><s>[INST] What color is the sky?[/INST] The sky is blue.</s>', '<s>[INST] What is the closest planet to the sun?[/INST] It is Mercury.</s>']
```

Having the special tokens can hinder the judge. This is particularly obvious with the llama:

```python
tokenizer = AutoTokenizer.from_pretrained("trl-internal-testing/tiny-random-Meta-Llama-3-8B-Instruct")
tokenizer.pad_token = tokenizer.eos_token
input_ids = tokenizer.apply_chat_template(messages, padding=True)

print(tokenizer.batch_decode(input_ids, skip_special_tokens=True))
['user\n\nWhat color is the sky?assistant\n\nThe sky is blue.', 'user\n\nWhat is the closest planet to the sun?assistant\n\nIt is Mercury.']
print(tokenizer.batch_decode(input_ids, skip_special_tokens=False))
['<|begin_of_text|><|start_header_id|>user<|end_header_id|>\n\nWhat color is the sky?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\nThe sky is blue.<|eot_id|><|eot_id|><|eot_id|>', '<|begin_of_text|><|start_header_id|>user<|end_header_id|>\n\nWhat is the closest planet to the sun?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\nIt is Mercury.<|eot_id|>']
```

File: trl/trainer/online_dpo_trainer.py [hunk 15]
```diff
@@ -421,23 +425,39 @@ def training_step(
...
+                prompt_completion_ids[:, context_length:], skip_special_tokens=True
+            )
+
+            ranks_of_first_completion = self.judge.judge(
+                prompts, [[completions[i], completions[i + num_examples]] for i in range(num_examples)]
+            )
+            # convert ranks to a True/False mask:
+            # when rank == 0, it means the first completion is the best
+            # when rank == 1, it means the second completion is the best
+            mask = torch.tensor([rank == 1 for rank in ranks_of_first_completion], device=prompt_completion_ids.device)
```
Comment (qgallouedec) [inline trl/trainer/online_dpo_trainer.py:443]: Can you double-check it's not this instead? It might be an error here
```suggestion
            mask = torch.tensor([rank == 0 for rank in ranks_of_first_completion], device=prompt_completion_ids.device)
```

Comment (kashif) [reply trl/trainer/online_dpo_trainer.py:443]: yes checking...

File: trl/trainer/xpo_trainer.py (hunk 14 above)
Comment (lewtun) [reply trl/trainer/xpo_trainer.py:227]: Yes, I would remove the padding tokens, but keep the special tokens. We have a helper method here if that's useful: https://github.com/huggingface/trl/blob/88be2c07e5024cc47e7777d23815602ae22c5f11/trl/trainer/utils.py#L1381

Comment (qgallouedec) [reply trl/trainer/xpo_trainer.py]: We've chosen another option, which is to retrieve the raw prompt (message list) and completion and format it with a simple chat template. Consequently, for a conversational dataset, regardless of the model's chat template, the judge will receive data in the following form:

```python
prompt = "User: What color is the sky\n\n"
completion = "Assistant: It is blue.\n\n"
```

see ce3798a

**Comment by qgallouedec** [comment]
https://wandb.ai/huggingface/huggingface/runs/jlff2gpp

```
accelerate launch examples/scripts/dpo_online.py \
    --model_name_or_path Qwen/Qwen2.5-0.5B-Instruct \
    --judge pair_rm \
    --dataset_name trl-lib/ultrafeedback-prompt \
    --learning_rate 5.0e-7 \
    --logging_steps 25 \
    --output_dir Qwen2.5-0.5B-Online-DPO-PairRM \
    --warmup_ratio 0.1
```

File: examples/scripts/dpo_online.py (hunk 12 above)
Comment (qgallouedec) [reply examples/scripts/dpo_online.py:88]: #2271

File: examples/scripts/dpo_online.py (hunk 13 above)
Comment (qgallouedec) [reply examples/scripts/dpo_online.py]: fbec78a
